// @filoz/synapse-sdk/dist/src/utils/constants.js
import { erc20Abi, multicall3Abi } from 'viem';
import { erc20PermitAbi } from "../abis/erc20-permit.js";
import * as abis from "../abis/gen.js";
export const TOKENS = {
    USDFC: 'USDFC',
    FIL: 'FIL',
};
export const CHAIN_IDS = {
    mainnet: 314,
    calibration: 314159,
};
export const CONTRACT_ABIS = {
    ERC20: erc20Abi,
    ERC20_PERMIT: erc20PermitAbi,
    PAYMENTS: abis.paymentsAbi,
    PDP_VERIFIER: abis.pdpVerifierAbi,
    WARM_STORAGE: abis.filecoinWarmStorageServiceAbi,
    WARM_STORAGE_VIEW: abis.filecoinWarmStorageServiceStateViewAbi,
    MULTICALL3: multicall3Abi,
    SERVICE_PROVIDER_REGISTRY: abis.serviceProviderRegistryAbi,
    SESSION_KEY_REGISTRY: abis.sessionKeyRegistryAbi,
};
export const EIP2612_PERMIT_TYPES = {
    Permit: [
        { name: 'owner', type: 'address' },
        { name: 'spender', type: 'address' },
        { name: 'value', type: 'uint256' },
        { name: 'nonce', type: 'uint256' },
        { name: 'deadline', type: 'uint256' },
    ],
};
export const TIME_CONSTANTS = {
    EPOCH_DURATION: 30,
    EPOCHS_PER_DAY: 2880n,
    EPOCHS_PER_MONTH: 86400n,
    DAYS_PER_MONTH: 30n,
    DEFAULT_LOCKUP_DAYS: 10n,
};
export const GENESIS_TIMESTAMPS = {
    mainnet: 1598306400,
    calibration: 1667326380,
};
export const SIZE_CONSTANTS = {
    KiB: 1024n,
    MiB: 1n << 20n,
    GiB: 1n << 30n,
    TiB: 1n << 40n,
    PiB: 1n << 50n,
    MAX_UPLOAD_SIZE: 200 * 1024 * 1024,
    MIN_UPLOAD_SIZE: 127,
    DEFAULT_UPLOAD_BATCH_SIZE: 32,
};
export const METADATA_KEYS = {
    WITH_CDN: 'withCDN',
    WITH_IPFS_INDEXING: 'withIPFSIndexing',
    IPFS_ROOT_CID: 'ipfsRootCID',
};
export const TIMING_CONSTANTS = {
    TRANSACTION_PROPAGATION_TIMEOUT_MS: 30000,
    TRANSACTION_PROPAGATION_POLL_INTERVAL_MS: 2000,
    DATA_SET_CREATION_TIMEOUT_MS: 7 * 60 * 1000,
    DATA_SET_CREATION_POLL_INTERVAL_MS: 2000,
    PIECE_PARKING_TIMEOUT_MS: 7 * 60 * 1000,
    PIECE_PARKING_POLL_INTERVAL_MS: 5000,
    TRANSACTION_CONFIRMATIONS: 1,
    PERMIT_DEADLINE_DURATION: 3600,
    PIECE_ADDITION_TIMEOUT_MS: 7 * 60 * 1000,
    PIECE_ADDITION_POLL_INTERVAL_MS: 1000,
};
export const SETTLEMENT_FEE = 1300000000000000n;
export const RPC_URLS = {
    mainnet: {
        http: 'https://api.node.glif.io/rpc/v1',
        websocket: 'wss://wss.node.glif.io/apigw/lotus/rpc/v1',
    },
    calibration: {
        http: 'https://api.calibration.node.glif.io/rpc/v1',
        websocket: 'wss://wss.calibration.node.glif.io/apigw/lotus/rpc/v1',
    },
};
export const CONTRACT_ADDRESSES = {
    WARM_STORAGE: {
        mainnet: '0x81DFD9813aDd354f03704F31419b0c6268d46232',
        calibration: '0x80617b65FD2EEa1D7fDe2B4F85977670690ed348',
    },
    MULTICALL3: {
        mainnet: '0xcA11bde05977b3631167028862bE2a173976CA11',
        calibration: '0xcA11bde05977b3631167028862bE2a173976CA11',
    },
    USDFC: {
        mainnet: '0x80B98d3aa09ffff255c3ba4A241111Ff1262F045',
        calibration: '0xb3042734b608a1B16e9e86B374A3f3e389B4cDf0',
    },
};
