// @filoz/synapse-sdk/dist/storage-service.js
import { MockUploadTask } from './upload-task.js';
import { asCommP } from './commp/index.js';
export class MockStorageService {
    proofSetId;
    storageProvider;
    _storedData = new Map();
    _signerAddress;
    _withCDN;
    constructor(proofSetId, storageProvider, signerAddress, withCDN) {
        this.proofSetId = proofSetId;
        this.storageProvider = storageProvider;
        this._signerAddress = signerAddress;
        this._withCDN = withCDN;
    }
    upload(data) {
        console.log('[MockSynapse] StorageService.upload() called');
        console.log('[MockSynapse] Data size:', data instanceof ArrayBuffer ? data.byteLength : data.length, 'bytes');
        const uploadTask = new MockUploadTask(data, this._withCDN);
        const storeData = async () => {
            console.log('[MockSynapse] Storing data internally for mock retrieval...');
            const commp = await uploadTask.commp();
            const bytes = data instanceof ArrayBuffer ? new Uint8Array(data) : data;
            this._storedData.set(commp.toString(), bytes);
            console.log('[MockSynapse] Data stored with CommP:', commp.toString());
        };
        storeData().catch(() => { });
        return uploadTask;
    }
    async download(commp, options) {
        const normalizedCommP = asCommP(commp);
        if (normalizedCommP == null) {
            throw new Error('Invalid CommP provided');
        }
        const commpString = normalizedCommP.toString();
        console.log('[MockSynapse] StorageService.download() called');
        console.log('[MockSynapse] CommP:', commpString);
        console.log('[MockSynapse] Download options:', options);
        if (options?.withCDN !== false && (this._withCDN || options?.withCDN)) {
            console.log('[MockSynapse] Using CDN for download (withCDN=true)');
            const res = await fetch(`https://${this._signerAddress}.calibration.filcdn.io/${commpString}`);
            return new Uint8Array(await res.arrayBuffer());
        }
        console.log('[MockSynapse] Simulating download network delay (300ms)...');
        await new Promise(resolve => setTimeout(resolve, 300));
        console.log('[MockSynapse] Looking up data in mock storage...');
        const data = this._storedData.get(commpString);
        if (data == null) {
            console.log('[MockSynapse] Data not found in mock storage!');
            throw new Error(`Data not found for CommP: ${commpString}`);
        }
        console.log('[MockSynapse] Data found, size:', data.length, 'bytes');
        if (options?.noVerify !== true) {
            console.log('[MockSynapse] Verifying data integrity...');
            console.log('[MockSynapse] Simulating verification delay (200ms)...');
            await new Promise(resolve => setTimeout(resolve, 200));
            console.log('[MockSynapse] Data verification complete');
        }
        else {
            console.log('[MockSynapse] Skipping verification (noVerify=true)');
        }
        console.log('[MockSynapse] Returning data copy');
        return new Uint8Array(data);
    }
    async delete(commp) {
        const normalizedCommP = asCommP(commp);
        if (normalizedCommP == null) {
            throw new Error('Invalid CommP provided');
        }
        const commpString = normalizedCommP.toString();
        console.log('[MockSynapse] StorageService.delete() called');
        console.log('[MockSynapse] CommP:', commpString);
        console.log('[MockSynapse] Simulating delete network delay (500ms)...');
        await new Promise(resolve => setTimeout(resolve, 500));
        console.log('[MockSynapse] Removing data from mock storage...');
        this._storedData.delete(commpString);
        console.log('[MockSynapse] Data deleted successfully');
    }
    async settlePayments() {
        console.log('[MockSynapse] StorageService.settlePayments() called');
        console.log('[MockSynapse] Simulating settlement processing delay (1000ms)...');
        await new Promise(resolve => setTimeout(resolve, 1000));
        const result = {
            settledAmount: 50000000000000000n,
            epoch: Math.floor(Date.now() / 30000)
        };
        console.log('[MockSynapse] Settlement complete:', result);
        return result;
    }
}
