// @filoz/synapse-sdk/dist/pdp/storage-provider.js
import { Contract } from 'ethers';
import { CONTRACT_ABIS } from '../utils/index.js';
export class StorageProviderTool {
    contract;
    signer;
    constructor(contractAddress, signer) {
        this.signer = signer;
        this.contract = new Contract(contractAddress, CONTRACT_ABIS.PANDORA_SERVICE, signer);
    }
    async register(pdpUrl, pieceRetrievalUrl) {
        return await this.contract.registerServiceProvider(pdpUrl, pieceRetrievalUrl);
    }
    async approve(providerAddress) {
        return await this.contract.approveServiceProvider(providerAddress);
    }
    async reject(providerAddress) {
        return await this.contract.rejectServiceProvider(providerAddress);
    }
    async remove(providerId) {
        return await this.contract.removeServiceProvider(providerId);
    }
    async isApproved(providerAddress) {
        return await this.contract.isProviderApproved(providerAddress);
    }
    async getProviderIdByAddress(providerAddress) {
        return await this.contract.getProviderIdByAddress(providerAddress);
    }
    async getApprovedProvider(providerId) {
        const result = await this.contract.getApprovedProvider(providerId);
        return {
            owner: result.owner,
            pdpUrl: result.pdpUrl,
            pieceRetrievalUrl: result.pieceRetrievalUrl,
            registeredAt: result.registeredAt,
            approvedAt: result.approvedAt
        };
    }
    async getPendingProvider(providerAddress) {
        const [pdpUrl, pieceRetrievalUrl, registeredAt] = await this.contract.pendingProviders(providerAddress);
        return {
            pdpUrl,
            pieceRetrievalUrl,
            registeredAt
        };
    }
    async getNextProviderId() {
        return await this.contract.nextServiceProviderId();
    }
    async getOwner() {
        return await this.contract.owner();
    }
    async getSignerAddress() {
        return await this.signer.getAddress();
    }
    async isOwner() {
        const [owner, signerAddress] = await Promise.all([
            this.getOwner(),
            this.getSignerAddress()
        ]);
        return owner.toLowerCase() === signerAddress.toLowerCase();
    }
    async getAllApprovedProviders() {
        const nextId = await this.getNextProviderId();
        const providers = [];
        for (let i = 1n; i < nextId; i++) {
            try {
                const info = await this.getApprovedProvider(i);
                if (info.owner !== '0x0000000000000000000000000000000000000000') {
                    providers.push({ id: i, info });
                }
            }
            catch (e) {
                continue;
            }
        }
        return providers;
    }
    getContractAddress() {
        return this.contract.target;
    }
}
