// @filoz/synapse-sdk/dist/pdp/tool.js
import { ethers } from 'ethers';
import { asCommP } from '../commp/index.js';
export class PDPTool {
    apiEndpoint;
    pdpAuthHelper;
    constructor(apiEndpoint, pdpAuthHelper) {
        if (apiEndpoint === '') {
            throw new Error('PDP API endpoint is required');
        }
        this.apiEndpoint = apiEndpoint.endsWith('/') ? apiEndpoint.slice(0, -1) : apiEndpoint;
        this.pdpAuthHelper = pdpAuthHelper;
    }
    async createProofSet(clientDataSetId, payee, withCDN, recordKeeper) {
        const authData = await this.pdpAuthHelper.signCreateProofSet(clientDataSetId, payee, withCDN);
        const extraData = this._encodeProofSetCreateData({
            metadata: '',
            payer: await this.pdpAuthHelper.getSignerAddress(),
            withCDN,
            signature: authData.signature
        });
        const requestBody = {
            recordKeeper,
            extraData: `0x${extraData}`
        };
        const response = await fetch(`${this.apiEndpoint}/pdp/proof-sets`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify(requestBody)
        });
        if (response.status !== 201) {
            const errorText = await response.text();
            throw new Error(`Failed to create proof set: ${response.status} ${response.statusText} - ${errorText}`);
        }
        const location = response.headers.get('Location');
        if (location == null) {
            throw new Error('Server did not provide Location header in response');
        }
        const locationMatch = location.match(/\/pdp\/proof-sets\/created\/(.+)$/);
        if (locationMatch == null) {
            throw new Error(`Invalid Location header format: ${location}`);
        }
        const txHash = locationMatch[1];
        return {
            txHash,
            statusUrl: `${this.apiEndpoint}${location}`
        };
    }
    async getProofSetCreationStatus(txHash) {
        const response = await fetch(`${this.apiEndpoint}/pdp/proof-sets/created/${txHash}`, {
            method: 'GET',
            headers: {
                'Content-Type': 'application/json'
            }
        });
        if (response.status === 404) {
            throw new Error(`Proof set creation not found for transaction hash: ${txHash}`);
        }
        if (response.status !== 200) {
            const errorText = await response.text();
            throw new Error(`Failed to get proof set creation status: ${response.status} ${response.statusText} - ${errorText}`);
        }
        return await response.json();
    }
    async addRoots(proofSetId, clientDataSetId, nextRootId, rootDataArray) {
        if (rootDataArray.length === 0) {
            throw new Error('At least one root must be provided');
        }
        for (const rootData of rootDataArray) {
            const commP = asCommP(rootData.cid);
            if (commP == null) {
                throw new Error(`Invalid CommP: ${String(rootData.cid)}`);
            }
        }
        const authData = await this.pdpAuthHelper.signAddRoots(clientDataSetId, nextRootId, rootDataArray);
        const extraData = this._encodeAddRootsExtraData({
            signature: authData.signature,
            metadata: ''
        });
        const requestBody = {
            roots: rootDataArray.map(rootData => {
                const cidString = typeof rootData.cid === 'string' ? rootData.cid : rootData.cid.toString();
                return {
                    rootCid: cidString,
                    subroots: [{
                            subrootCid: cidString
                        }]
                };
            }),
            extraData: `0x${extraData}`
        };
        const response = await fetch(`${this.apiEndpoint}/pdp/proof-sets/${proofSetId}/roots`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify(requestBody)
        });
        if (response.status !== 201) {
            const errorText = await response.text();
            throw new Error(`Failed to add roots to proof set: ${response.status} ${response.statusText} - ${errorText}`);
        }
        const responseText = await response.text();
        return {
            message: responseText !== '' ? responseText : `Roots added to proof set ID ${proofSetId} successfully`
        };
    }
    getApiEndpoint() {
        return this.apiEndpoint;
    }
    getPDPAuthHelper() {
        return this.pdpAuthHelper;
    }
    _encodeProofSetCreateData(data) {
        const signature = data.signature.startsWith('0x') ? data.signature : `0x${data.signature}`;
        const abiCoder = ethers.AbiCoder.defaultAbiCoder();
        const encoded = abiCoder.encode(['string', 'address', 'bool', 'bytes'], [data.metadata, data.payer, data.withCDN, signature]);
        return encoded.slice(2);
    }
    _encodeAddRootsExtraData(data) {
        const signature = data.signature.startsWith('0x') ? data.signature : `0x${data.signature}`;
        const abiCoder = ethers.AbiCoder.defaultAbiCoder();
        const encoded = abiCoder.encode(['bytes', 'string'], [signature, data.metadata]);
        return encoded.slice(2);
    }
}
